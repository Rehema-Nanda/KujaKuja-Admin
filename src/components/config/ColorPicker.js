import React from "react";
import { SketchPicker } from "react-color";
import reactCSS from "reactcss";
import PropTypes from "prop-types";

import "../Dashboard.css";
import "../../index.css";
import { Col, Row } from "reactstrap";

class ColorPicker extends React.Component {
    // https://casesandberg.github.io/react-color/
    // https://github.com/casesandberg/reactcss

    static get DEFAULT_COLOR() {
        return "#000";
    }

    constructor(props) {
        super(props);

        this.state = {
            displayColorPicker: false,
            color: ColorPicker.DEFAULT_COLOR,
        };
    }

    componentDidUpdate(props) {
        const { refresh, color } = this.props;
        if (props.refresh !== refresh) {
            // eslint-disable-next-line react/no-did-update-set-state
            this.setState({
                color: color,
            });
        }
    }


    handleClick = () => {
        const { displayColorPicker } = this.state;

        this.setState({ displayColorPicker: !displayColorPicker });
    };

    handleClose = () => {
        this.setState({ displayColorPicker: false });
    };

    handleChange = (color) => {
        const { handleColorChange } = this.props;

        this.setState({ color: color.hex }, () => {
            handleColorChange(color);
        });
    };

    render() {
        const { color, displayColorPicker } = this.state;
        const styles = reactCSS({
            default: {
                color: {
                    width: "60px",
                    height: "60px",
                    borderRadius: "2px",
                    background: color,
                },
                swatch: {
                    padding: "4px",
                    background: "#fff",
                    borderRadius: "1px",
                    boxShadow: "0 0 0 1px rgba(0,0,0,.1)",
                    display: "inline-block",
                    cursor: "pointer",
                },
                popover: {
                    position: "absolute",
                    zIndex: "2",
                },
                cover: {
                    position: "fixed",
                    top: "0px",
                    right: "0px",
                    bottom: "0px",
                    left: "0px",
                },
                picker_col: {
                    textAlign: "center",
                },
            },
        });

        return (
            <Row>
                <Col style={styles.picker_col} sm="12" md="6">
                    <div style={styles.swatch} onClick={this.handleClick} role="presentation">
                        <div style={styles.color} />
                    </div>
                    {displayColorPicker
                        ? (
                            <div style={styles.popover}>
                                <div style={styles.cover} onClick={this.handleClose} role="presentation" />
                                <SketchPicker id="sketch" color={color || ColorPicker.DEFAULT_COLOR} onChange={this.handleChange} />
                            </div>
                        ) : null}
                </Col>
            </Row>
        );
    }
}

ColorPicker.propTypes = {
    refresh: PropTypes.bool.isRequired,
    color: PropTypes.string.isRequired,
    handleColorChange: PropTypes.func.isRequired,
};

export default ColorPicker;
