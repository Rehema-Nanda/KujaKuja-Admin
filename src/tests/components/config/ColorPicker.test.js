import React from "react";
import { shallow } from "enzyme";
import ColorPicker from "../../../components/config/ColorPicker";

describe("<ColorPicker />", () => {
    const testProps = {
        handleColorChange: jest.fn(),
        color: "blue",
    };
    const wrapper = shallow(<ColorPicker {...testProps} />);

    it("should render correctly", () => {
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it("should call handleClick function when the div is clicked", () => {
        const testFunc = jest.spyOn(wrapper.instance(), "handleClick");
        wrapper.instance().forceUpdate(); // reason here https://github.com/airbnb/enzyme/issues/944
        const div = wrapper.find("div").first();
        div.simulate("click");
        expect(testFunc).toHaveBeenCalled();
    });

    it("should call handleChange function when the color changes", () => {
        const testFunc = jest.spyOn(wrapper.instance(), "handleChange");
        wrapper.setState({ displayColorPicker: true });
        const color = {
            hex: "",
        };
        expect(wrapper.state().displayColorPicker).toEqual(true);
        const sketchPicker = wrapper.find("#sketch");
        sketchPicker.simulate("change", color);
        wrapper.instance().forceUpdate(); // reason here https://github.com/airbnb/enzyme/issues/944
        expect(testFunc).toHaveBeenCalled();
    });
    it("updates the state color after the prop changes", () => {
        wrapper.setProps({ refresh: "something" });
        expect(wrapper.state().color).toEqual("blue");
    });

    it("should set the displayColorPicker state to false when handleClose is called ", () => {
        const div = wrapper.find("div").at(3);
        div.simulate("click");
        expect(wrapper.state().displayColorPicker).toEqual(false);
    });
});
