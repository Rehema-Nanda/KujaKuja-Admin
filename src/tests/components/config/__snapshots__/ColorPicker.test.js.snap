// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorPicker /> should render correctly 1`] = `
"<Row tag=\\"div\\">
  <Col style={{...}} sm=\\"12\\" md=\\"6\\" tag=\\"div\\" widths={{...}}>
    <div style={{...}} onClick={[Function]} role=\\"presentation\\">
      <div style={{...}} />
    </div>
  </Col>
</Row>"
`;
